firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexFreeQueryEngine.java

```diff
@@ -0,0 +1,78 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.local;
+
+import com.google.firebase.database.collection.ImmutableSortedMap;
+import com.google.firebase.database.collection.ImmutableSortedSet;
+import com.google.firebase.firestore.core.Query;
+import com.google.firebase.firestore.model.Document;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.MaybeDocument;
+import com.google.firebase.firestore.model.SnapshotVersion;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+public class IndexFreeQueryEngine implements QueryEngine {
+  private final LocalDocumentsView localDocumentsView;
+  private final QueryCache queryCache;
+
+  public IndexFreeQueryEngine(LocalDocumentsView localDocumentsView, QueryCache queryCache) {
+    this.localDocumentsView = localDocumentsView;
+    this.queryCache = queryCache;
+  }
+
+  @Override
+  public ImmutableSortedMap<DocumentKey, Document> getDocumentsMatchingQuery(
```

Comment:
Note for future me: We should probably update this to parse in the set of RemoteKeys as we are otherwise fetching it twice (once in executeQuery, once here).